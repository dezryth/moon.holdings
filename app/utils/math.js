// Internal functions
const floor = num => (Math.floor(num * 100) / 100);

// Exported
export const round = value => Math.round((value) * 100) / 100;

export const multiply = (num1, num2) => num1 * num2;

export const rounder = (balance, priceUsd) => round(multiply(balance, floor(priceUsd)));

export const numberWithCommas = x => x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

export const calcTotal = (assets) => {
  const values = assets.map(asset => asset.value);
  const total = values.reduce((sum, value) => sum + value, 0);
  return numberWithCommas(floor(total));
};

export const calculateBalance = coin => numberWithCommas(floor(coin.balance * coin.price_usd));

export const getCoinValue = coin => coin.value;

export const portfolioBalance = coins =>
  numberWithCommas(floor(coins.reduce((val, coin) => (val + +(coin.value)), 0)));
