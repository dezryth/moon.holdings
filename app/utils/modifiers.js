/* eslint-disable no-param-reassign */
/* eslint-disable no-extend-native */
/* eslint-disable func-names */
/* eslint-disable no-useless-escape */
import * as R from 'ramda';

// Styles
import * as style from 'styles/coins';

// Converts snake-case to camelCase.
String.prototype.toCamel = function () {
  return this.replace(/(\-[a-z])/g, $1 => $1.toUpperCase().replace('-', ''));
};

// Add style to coin square.
export const setStyle = (id) => {
  switch (id) {
    case 'all-sports': return style.basicattentiontoken;
    case 'basic-attention-token': return style.basicattentiontoken;
    case 'binance-coin':
    case 'bitcoin-cash':
    case 'bibox-token':
    case 'ethereum-classic':
    case 'deepbrain-chain':
    case 'golem-network-tokens':
    case 'gnosis-gno':
    case 'enjin-coin':
    case 'kucoin-shares':
    case 'kyber-network':
    case 'matrix-ai-network':
    case 'theta-token': return style[id.toCamel()];
    case '0x': return style.zrx;
    default: return style[id];
  }
};

// Check it coin has darkBg.
export const coinHasDarkBg = id =>
  (R.isNil(style[id.toCamel()]) ? false : style[id.toCamel()].darkBg);

// To handle special cases.
const coinIdChecker = (id) => {
  switch (id) {
    case '0x': return 'zrx';
    default: return id;
  }
};

// Check it coin has lightBg.
export const coinHasLightBg = (id) => {
  const coinId = coinIdChecker(id);
  return (R.isNil(style[coinId.toCamel()]) ? false : style[coinId.toCamel()].lightBg);
};

// Convert Array to Object.
export const arrayToObject = array =>
  array.reduce((obj, item) => {
    obj[item.id] = item;
    return obj;
  }, {});

const coinBg = () => 'coin-square bg';

// Add special Coin style.
export const classModifier = (coinId) => {
  switch (coinId) {
    case 'bitcoin': return `${coinBg()} bg-btc`;
    case 'ethereum': return `${coinBg()} bg-eth`;
    case 'eos': return `${coinBg()} bg-eos`;
    case 'lisk': return `${coinBg()} bg-lsk`;
    case 'neo': return `${coinBg()} bg-neo`;
    case 'zencash': return `${coinBg()} bg-zen`;
    default: // no defualt
  }
  return 'coin-square';
};

export const styleModifier = coinId => (coinHasDarkBg(coinId) ? `col ${classModifier(coinId)} dark-bg` : `col ${classModifier(coinId)} coin-square`);
