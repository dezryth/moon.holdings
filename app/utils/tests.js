import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import toJson from 'enzyme-to-json';

configure({ adapter: new Adapter() });

/**
 * Automatically tests that the component matches the Jest snapshot.
 * NOTE: If you are receiving a warning like the following in your tests:
 *
 *   Warning: React.createElement: type is invalid -- expected a string...
 *
 * Then you most likely need to pass the appropriate props.
 *
 * @param {!React.Component} Component The React component to wrap and test.
 * @param {!Object} props The (optional) props to use for the basic Jest test.
 */
export const testCommonComponentAttrs = (Component, props) => {
  describe('when rendering', () => {
    const wrapper = shallow(<Component {...props} />);

    it('should render a component matching the snapshot', () => {
      const tree = toJson(wrapper);
      expect(tree).toMatchSnapshot();
      expect(wrapper).toHaveLength(1);
    });
  });
};

export const getComponentWrapper = (Component, props) =>
  shallow(<Component {...props} />);
