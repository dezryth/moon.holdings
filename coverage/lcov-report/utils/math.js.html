<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for utils/math.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../index.html">All files</a> / <a href="index.html">utils</a> math.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">47.62% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>10/21</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">18.18% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>2/11</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">66.67% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>8/12</span>
      </div>
    </div>
  </div>
  <div class='status-line low'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">// Internal functions
const floor = num =&gt; (Math.floor(num * 100) / 100);
&nbsp;
// Exported
export const round = <span class="fstat-no" title="function not covered" >va</span>lue =&gt; <span class="cstat-no" title="statement not covered" >Math.round((value) * 100) / 100;</span>
&nbsp;
export const multiply = <span class="fstat-no" title="function not covered" >(n</span>um1, num2) =&gt; <span class="cstat-no" title="statement not covered" >num1 * num2;</span>
&nbsp;
export const rounder = <span class="fstat-no" title="function not covered" >(b</span>alance, priceUsd) =&gt; <span class="cstat-no" title="statement not covered" >round(multiply(balance, floor(priceUsd)));</span>
&nbsp;
export const numberWithCommas = <span class="fstat-no" title="function not covered" >x </span>=&gt; <span class="cstat-no" title="statement not covered" >x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');</span>
&nbsp;
export const calcTotal = <span class="fstat-no" title="function not covered" >(a</span>ssets) =&gt; {
  const values = <span class="cstat-no" title="statement not covered" >assets.map(<span class="fstat-no" title="function not covered" >as</span>set =&gt; <span class="cstat-no" title="statement not covered" >asset.value)</span>;</span>
  const total = <span class="cstat-no" title="statement not covered" >values.reduce(<span class="fstat-no" title="function not covered" >(s</span>um, value) =&gt; <span class="cstat-no" title="statement not covered" >sum + value,</span> 0);</span>
<span class="cstat-no" title="statement not covered" >  return numberWithCommas(floor(total));</span>
};
&nbsp;
export const calculateBalance = coin =&gt; floor(coin.balance * coin.price_usd);
&nbsp;
export const portfolioBalance = <span class="fstat-no" title="function not covered" >co</span>ins =&gt;
<span class="cstat-no" title="statement not covered" >  numberWithCommas(floor(coins.reduce(<span class="fstat-no" title="function not covered" >(v</span>al, coin) =&gt; (<span class="cstat-no" title="statement not covered" >val + +(coin.value))</span>, 0)));</span>
&nbsp;
// export const formatPercentage = (percent) =&gt; {
//   const num = percent || 0;
//   const currency = num.toFixed(2);
//   const split = currency.split('.');
//   const interger = parseInt(split[0]);
//   const decimals = parseInt(split[1]);
//
//   if (decimals === 0) {
//     interger.toFixed(0);
//     return interger;
//   }
//
//   return currency;
// };
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sat May 12 2018 16:09:30 GMT-0500 (CDT)
</div>
</div>
<script src="../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../sorter.js"></script>
</body>
</html>
