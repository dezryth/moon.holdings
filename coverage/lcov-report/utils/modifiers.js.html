<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for utils/modifiers.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../index.html">All files</a> / <a href="index.html">utils</a> modifiers.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">46.15% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>6/13</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">33.33% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>1/3</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">20% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>1/5</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">50% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>6/12</span>
      </div>
    </div>
  </div>
  <div class='status-line low'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">/* eslint-disable no-param-reassign */
/* eslint-disable no-extend-native */
/* eslint-disable func-names */
/* eslint-disable no-useless-escape */
&nbsp;
// Styles
import * as style from 'styles/coins';
&nbsp;
// Converts snake-case to camelCase
String.prototype.toCamel = <span class="fstat-no" title="function not covered" >fu</span>nction () {
<span class="cstat-no" title="statement not covered" >  return this.replace(/(\-[a-z])/g, <span class="fstat-no" title="function not covered" >$1</span> =&gt; <span class="cstat-no" title="statement not covered" >$1.toUpperCase().replace('-', ''))</span>;</span>
};
&nbsp;
export const formatPriceUSD = <span class="fstat-no" title="function not covered" >(c</span>oin) =&gt; {
<span class="cstat-no" title="statement not covered" >  coin.price_usd = Number(coin.price_usd).toFixed(2);</span>
<span class="cstat-no" title="statement not covered" >  return coin;</span>
};
&nbsp;
export const zeroBalanceValue = <span class="fstat-no" title="function not covered" >co</span>in =&gt;
<span class="cstat-no" title="statement not covered" >  Object.assign({</span>
    balance: 0,
    value: 0,
    percentage: 0
  }, coin);
&nbsp;
export const setStyle = (id) =&gt; {
  switch (id) {
<span class="branch-0 cbranch-no" title="branch not covered" >    case 'bitcoin-cash': <span class="cstat-no" title="statement not covered" >return style[id.toCamel()];</span></span>
<span class="branch-1 cbranch-no" title="branch not covered" >    case '0x': <span class="cstat-no" title="statement not covered" >return style.zrx;</span></span>
    default: return style[id];
  }
};
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sat May 12 2018 16:09:30 GMT-0500 (CDT)
</div>
</div>
<script src="../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../sorter.js"></script>
</body>
</html>
