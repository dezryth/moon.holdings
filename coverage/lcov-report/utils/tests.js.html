<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for utils/tests.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../index.html">All files</a> / <a href="index.html">utils</a> tests.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">90% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>9/10</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">75% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>3/4</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">90% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>9/10</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
&nbsp;
configure({ adapter: new Adapter() });
&nbsp;
/**
 * Automatically tests that the component matches the Jest snapshot.
 * NOTE: If you are receiving a warning like the following in your tests:
 *
 *   Warning: React.createElement: type is invalid -- expected a string...
 *
 * Then you most likely need to pass the appropriate props.
 *
 * @param {!React.Component} Component The React component to wrap and test.
 * @param {!Object} props The (optional) props to use for the basic Jest test.
 */
export const testCommonComponentAttrs = (Component, props) =&gt; {
  describe('when rendering', () =&gt; {
    const wrapper = shallow(&lt;Component {...props} /&gt;);
&nbsp;
    it('should render a component matching the snapshot', () =&gt; {
      const tree = toJson(wrapper);
      expect(tree).toMatchSnapshot();
      expect(wrapper).toHaveLength(1);
    });
  });
};
&nbsp;
export const getComponentWrapper = <span class="fstat-no" title="function not covered" >(C</span>omponent, props) =&gt;
<span class="cstat-no" title="statement not covered" >  shallow(&lt;Component {...props} /&gt;);</span>
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sat May 12 2018 16:09:30 GMT-0500 (CDT)
</div>
</div>
<script src="../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../sorter.js"></script>
</body>
</html>
