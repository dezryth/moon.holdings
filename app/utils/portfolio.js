/* eslint-disable no-param-reassign */
/* eslint-disable prefer-destructuring */

// Libs
import * as R from 'ramda';

// Utils
import { round } from 'utils/math';

const { localStorage } = window;

// Convert Array to Object
const arrayToObject = array =>
  array.reduce((obj, item) => {
    obj[item.id] = item;
    return obj;
  }, {});

// Add coin's percentage of portfolio
export const calculatePercentage = (portfolio, coin) => {
  portfolio.push(coin);

  const addValue = c => c.value;
  const values = R.chain(addValue, portfolio);
  const total = values.reduce((acc, val) => acc + val);

  const updatedPortfolio = portfolio.map((c) => {
    c.percentage = round((c.value / total) * 100);
    return c;
  });

  const moonPortfolio = arrayToObject(updatedPortfolio);

  localStorage.setItem('moonPortfolio', JSON.stringify(moonPortfolio));

  return updatedPortfolio;
};
