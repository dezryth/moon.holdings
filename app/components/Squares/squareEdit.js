/* eslint-disable class-methods-use-this */
/* eslint-disable no-unused-expressions */
import React from 'react';
import { connect } from 'react-redux';

// Actions
import { addCoin, updateCoin, removeCoin } from 'actions/coins';

// Utils
import { round, rounder } from 'utils/math';
import { setStyle } from 'utils/modifiers';

// Styles
import * as style from 'styles/coins';

class SquareEdit extends React.Component {
  constructor(props) {
    super(props);

    const { coin, coins } = this.props;
    const { portfolio } = coins;
    const { price_usd: price } = coin;

    this.state = {
      coin,
      price,
      balance: 0,
      value: '',
      portfolio,
      inPortfolio: false
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.handleSave = this.handleSave.bind(this);
    this.handleRemove = this.handleRemove.bind(this);
  }

  componentWillMount() {
    const { coin, balance: stateBalance } = this.state;
    const { portfolio } = this.state;
    const inPortfolio = portfolio.filter(c => c.id === coin.id);
    const portCoin = inPortfolio[0] ? inPortfolio[0] : coin;
    const balance = portCoin.balance ? portCoin.balance : stateBalance;

    this.setState({
      coin: portCoin,
      balance,
      inPortfolio: inPortfolio.length > 0
    });
  }

  handleFocus(e) {
    e.target.select();
  }

  handleChange(event) {
    const balance = event.target.value;
    const value = rounder(balance, this.state.price);

    this.setState({
      balance,
      value
    });
  }

  addCoin(coin, balance) {
    this.props.addCoin(Object.assign({ balance }, coin));
  }

  updateCoin(coin, balance) {
    const updatedCoin = Object.assign(coin);
    updatedCoin.balance = balance;
    this.props.updateCoin(updatedCoin);
  }

  handleSave() {
    const { coin, balance, inPortfolio } = this.state;
    inPortfolio ? this.updateCoin(coin, balance) : this.addCoin(coin, balance);
    this.props.closeEdit();
  }

  handleRemove() {
    const { coin } = this.state;
    this.props.removeCoin(coin);
    this.props.closeEdit();
  }

  squareHeight() {
    return this.state.inPortfolio ? 'edit-tall' : 'edit-short';
  }

  renderRemoveButton() {
    return (
      <button id="save-button" onClick={this.handleRemove}>
        Remove
      </button>
    );
  }

  render() {
    const { coin, value, inPortfolio } = this.state;
    const { symbol, price_usd: price } = coin;
    const isDisabled = this.state.balance <= 0;

    return (
      <div id="square-edit-container">
        <button className="close-modal-x" onClick={this.props.closeEdit} />
        <section
          id="square-edit"
          className={this.squareHeight()}
          style={setStyle(coin.id, style)}
        >
          <input
            id="coin-balance"
            type="number"
            placeholder="0"
            value={this.state.balance}
            onFocus={this.handleFocus}
            onChange={this.handleChange}
          />
          <div className="symbol">
            {symbol}
          </div>
          <div id="edit-coin-price">
            1 @ ${round(price)}
          </div>
          <div id="edit-total-value">
            ${value}
          </div>
          <button id="save-button" onClick={this.handleSave} disabled={isDisabled}>
            Save
          </button>
          { inPortfolio ? this.renderRemoveButton() : null }
        </section>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addCoin: (...args) => dispatch(addCoin(...args)),
  updateCoin: (...args) => dispatch(updateCoin(...args)),
  removeCoin: (...args) => dispatch(removeCoin(...args))
});

const mapStateToProps = ({ coins }) => ({
  coins
});

export const SquareEditJest = SquareEdit;

export default connect(mapStateToProps, mapDispatchToProps)(SquareEdit);
