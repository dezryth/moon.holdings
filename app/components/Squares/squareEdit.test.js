import { testCommonComponentAttrs } from '../../utils/tests';

import { SquareEditJest } from './squareEdit';

const coin = {
  id: 'btc',
  symbol: 'bitcoin',
  balance: 0,
  price_usd: '1'
};

const coins = {
  all: [coin],
  portfolio: []
};

const props = {
  coins,
  coin,
  addCoin: jest.fn(),
  updateCoin: jest.fn(),
  removeCoin: jest.fn()
};

describe('<SquareEditJest /> component', () => {
  testCommonComponentAttrs(SquareEditJest, { ...props });
});
