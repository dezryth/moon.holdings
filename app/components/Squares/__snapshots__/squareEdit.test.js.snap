// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SquareEditJest /> component when rendering should render a component matching the snapshot 1`] = `
<div
  id="square-edit-container"
>
  <button
    className="close-modal-x"
  />
  <section
    className="edit-short"
    id="square-edit"
  >
    <input
      id="coin-balance"
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="0"
      type="number"
      value={0}
    />
    <div
      className="symbol"
    >
      bitcoin
    </div>
    <div
      id="edit-coin-price"
    >
      1 @ $
      1
    </div>
    <div
      id="edit-total-value"
    >
      $
    </div>
    <button
      disabled={true}
      id="save-button"
      onClick={[Function]}
    >
      Save
    </button>
  </section>
</div>
`;
